Improve performance of immutable maps

2opremio opened the pull request:
* Replace github.com/mndrix/ps by github.com/weaveworks/ps (See https://github.com/weaveworks/ps/pull/1 )
* LatestMap codec performance improvements and cleanups 
   * Allocate all map entries of the intermediate representation at once
   * Use UnsafeMutableMap to improve performance of LatestMap construction
   * Remove gob encoder/decoder

Improves #1010 #1457  #971

tomwilkie commented:
I like that `UnsafeMutableSet` is on the ps.Tree, and not on the LatestMap.

What is the performance impact?

2opremio commented:
My measurements against the query service in the `dev-c4` cluster show over a ~10% reduction in CPU consumption. (For @rade , I initially saw about 30% but stabilized in ~10% when I removed load balancing from the equation, using single replica, and I obtained longer profiles of 90s instead of 30s)

Without the changes in this PR:

```
$ go tool pprof --seconds 90 http://localhost:4040/debug/pprof/profile
Fetching profile from http://localhost:4040/debug/pprof/profile?seconds=90
Please wait... (1m30s)
Saved profile in /Users/fons/pprof/pprof.localhost:4040.samples.cpu.001.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top10
21.96s of 41.59s total (52.80%)
Dropped 382 nodes (cum <= 0.21s)
Showing top 10 nodes out of 197 (cum >= 1.13s)
      flat  flat%   sum%        cum   cum%
     5.12s 12.31% 12.31%      8.14s 19.57%  runtime.scanobject
     3.32s  7.98% 20.29%      6.26s 15.05%  runtime.heapBitsSweepSpan
     2.94s  7.07% 27.36%      2.94s  7.07%  runtime.(*mspan).sweep.func1
     2.30s  5.53% 32.89%      8.21s 19.74%  runtime.mallocgc
     1.96s  4.71% 37.61%      1.98s  4.76%  runtime.heapBitsSetType
     1.57s  3.77% 41.38%      1.57s  3.77%  runtime.memmove
     1.24s  2.98% 44.36%      1.94s  4.66%  runtime.greyobject
     1.23s  2.96% 47.32%      2.38s  5.72%  github.com/weaveworks/scope/vendor/github.com/mndrix/ps.hashKey
     1.15s  2.77% 50.08%      1.15s  2.77%  runtime.stringiter2
     1.13s  2.72% 52.80%      1.13s  2.72%  runtime.heapBitsForObject
(pprof) 
```

With the changes:

```
$ go tool pprof --seconds 90 http://localhost:4040/debug/pprof/profile
Fetching profile from http://localhost:4040/debug/pprof/profile?seconds=90
Please wait... (1m30s)
Saved profile in /Users/fons/pprof/pprof.localhost:4040.samples.cpu.002.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top10
20.28s of 38.11s total (53.21%)
Dropped 409 nodes (cum <= 0.19s)
Showing top 10 nodes out of 200 (cum >= 1.01s)
      flat  flat%   sum%        cum   cum%
     4.91s 12.88% 12.88%      8.36s 21.94%  runtime.scanobject
     3.38s  8.87% 21.75%      5.54s 14.54%  runtime.heapBitsSweepSpan
     2.16s  5.67% 27.42%      2.16s  5.67%  runtime.(*mspan).sweep.func1
     1.93s  5.06% 32.48%      6.01s 15.77%  runtime.mallocgc
     1.53s  4.01% 36.50%      1.53s  4.01%  runtime.memmove
     1.44s  3.78% 40.28%      1.44s  3.78%  runtime.heapBitsSetType
     1.38s  3.62% 43.90%      2.15s  5.64%  runtime.greyobject
     1.33s  3.49% 47.39%      1.33s  3.49%  runtime.heapBitsForObject
     1.21s  3.18% 50.56%      2.22s  5.83%  github.com/weaveworks/scope/vendor/github.com/mndrix/ps.hashKey
     1.01s  2.65% 53.21%      1.01s  2.65%  runtime.stringiter2
(pprof) 
```

2opremio commented:
After fixing the cut-and-paste bug in the recursive call things look better: ~18% improvement:

```
$ go tool pprof --seconds 90 http://localhost:4040/debug/pprof/profile
Fetching profile from http://localhost:4040/debug/pprof/profile?seconds=90
Please wait... (1m30s)
Saved profile in /Users/fons/pprof/pprof.localhost:4040.samples.cpu.005.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top10
17710ms of 34480ms total (51.36%)
Dropped 400 nodes (cum <= 172.40ms)
Showing top 10 nodes out of 198 (cum >= 970ms)
      flat  flat%   sum%        cum   cum%
    3500ms 10.15% 10.15%     6040ms 17.52%  runtime.scanobject
    2710ms  7.86% 18.01%     2710ms  7.86%  runtime.(*mspan).sweep.func1
    2520ms  7.31% 25.32%     5240ms 15.20%  runtime.heapBitsSweepSpan
    1620ms  4.70% 30.02%     6440ms 18.68%  runtime.mallocgc
    1550ms  4.50% 34.51%     1550ms  4.50%  runtime.memmove
    1340ms  3.89% 38.40%     2450ms  7.11%  github.com/weaveworks/scope/vendor/github.com/mndrix/ps.hashKey
    1320ms  3.83% 42.23%     1320ms  3.83%  runtime.heapBitsSetType
    1110ms  3.22% 45.45%     1110ms  3.22%  runtime.stringiter2
    1070ms  3.10% 48.55%     1620ms  4.70%  runtime.greyobject
     970ms  2.81% 51.36%      970ms  2.81%  runtime.heapBitsForObject
(pprof) 
```

tomwilkie commented:
Thats a bit better!

2opremio commented:
I've discovered that we are spending 10% of our time ... parsing unicode while hashing

```
func hashKey(key string) uint64 {
	hash := offset64
	for _, codetime := range key {
		hash ^= uint64(codepoint)
		hash *= prime64
	}
	return hash
}
```

An unsafe casting to a bytes cuts down another 10% CPU:

```
$ go tool pprof --seconds 90 http://localhost:4040/debug/pprof/profile
Fetching profile from http://localhost:4040/debug/pprof/profile?seconds=90
Please wait... (1m30s)
Saved profile in /Users/fons/pprof/pprof.localhost:4040.samples.cpu.006.pb.gz
Entering interactive mode (type "help" for commands) 
(pprof) top5
11.88s of 31.75s total (37.42%)
Dropped 374 nodes (cum <= 0.16s)
Showing top 5 nodes out of 206 (cum >= 1.29s)
      flat  flat%   sum%        cum   cum%
     3.58s 11.28% 11.28%      6.09s 19.18%  runtime.scanobject
     3.06s  9.64% 20.91%      5.34s 16.82%  runtime.heapBitsSweepSpan
     2.28s  7.18% 28.09%      2.28s  7.18%  runtime.(*mspan).sweep.func1
     1.67s  5.26% 33.35%      6.12s 19.28%  runtime.mallocgc
     1.29s  4.06% 37.42%      1.29s  4.06%  runtime.memmove
(pprof) 
```

tomwilkie commented on common/xfer/plugin_spec.go:
```diff
@@ -7,8 +7,8 @@ import (
 	"github.com/davecgh/go-spew/spew"
-	"github.com/mndrix/ps"
 	"github.com/ugorji/go/codec"
+	"github.com/weaveworks/ps"
```
Generally we don't change the imports when forking something, we just checkout the fork in the original vendored location.

2opremio replied:
> Generally we don't change the imports when forking something, we just checkout the fork in the original vendored location.

Where have we done this? (honest question). I think it's a bad practice since it's misleading on what libraries you are really using. Also, it may conflict with the gvt tool.

tomwilkie replied:
gvt works nicely for this.

We did this multiple times for the docker client library for instance.  It was always out intention to upstream the changes though, which may be different here?  Have you tried asking the original author?

2opremio replied:
The library hasn't been updated for 4 years, I doubt there's any interest in upstreaming the changes, although I could ask. In the meantime I think we should be using our fork explicitly.

2opremio replied:
Done https://github.com/mndrix/ps/pull/5

I commit to replace the vendoring by the upstream version if the changes get merged.

tomwilkie replied:
In this case you're lucky, as its a single package library that doesn't seem to reference itself.  Other more complicated libraries have references to themselves in their code, and cannot just be checked out in arbitrary paths.

We should be running on our own fork for a long time.  If the upstream author doesn't want to take the change, we should work with him to find a change that is suitable to both of us.  If this fails, we have to take on the maintenance of our fork (and for instance implement CI etc).

tomwilkie commented:
Other than one comment, LGTM

paulbellamy commented on report/latest_map.go:
```diff
@@ -178,14 +177,6 @@ func (m LatestMap) toIntermediate() map[string]LatestEntry {
-
 // CodecEncodeSelf implements codec.Selfer
 func (m *LatestMap) CodecEncodeSelf(encoder *codec.Encoder) {
 	if m.Map != nil {
```
I'm assuming we could use `codec.GenHelperEncoer(encoder)` here to do the same thing?

2opremio replied:
Yep, I think that @tomwilkie just didn't get around to it in #1709

tomwilkie replied:
I didn't do it to ensure we were backwards compatible.

2opremio replied:
I thought we could do the encoding in a backwards compatible way. Can't we?

paulbellamy replied:
Discussed elsewhere, but not doing the reader *now* means that we can't *accidentally* break backwards-compatibility.

tomwilkie replied:
> I thought we could do the encoding in a backwards compatible way. Can't we?

We can, but then its gets harder to detect when we break backwards compatibility.

paulbellamy commented:
You probably meant to `gvt fetch github.com/weaveworks/ps`

2opremio commented:
> You probably meant to `gvt fetch github.com/weaveworks/ps`

I forgot to commit it, thanks

paulbellamy commented:
Once vendored, LGTM.
